agent/asm/asm_test.go

```diff
@@ -111,11 +121,74 @@ func TestASMGetAuthConfig(t *testing.T) {
 }
 
 func TestGetSecretFromASM(t *testing.T) {
-	asmClient := mockGetSecretValue{
-		Resp: secretsmanager.GetSecretValueOutput{
-			SecretString: aws.String("secretValue"),
-		},
-	}
+	asmClient := createASMInterface(secretValue)
 	_, err := GetSecretFromASM("secretName", asmClient)
 	assert.NoError(t, err)
 }
+
+func TestGetSecretFromASMWithJsonKey(t *testing.T) {
+	asmClient := createASMInterface(jsonSecretValue)
+	secretValueInput := createSecretValueInput(toPtr(valueFrom), nil, nil)
+	outSecretValue, _ := GetSecretFromASMWithInput(secretValueInput, asmClient, jsonKey)
+	assert.Equal(t, secretValue, outSecretValue)
+}
+
+func TestGetSecretFromASMWithMalformedJSON(t *testing.T) {
+	asmClient := createASMInterface(malformedJsonSecretValue)
+	secretValueInput := createSecretValueInput(toPtr(valueFrom), nil, nil)
+	outSecretValue, err := GetSecretFromASMWithInput(secretValueInput, asmClient, jsonKey)
+	if assert.Error(t, err) {
+		assert.Equal(t, "", outSecretValue)
+	}
+}
+
+func TestGetSecretFromASMWithJSONKeyNotFound(t *testing.T) {
+	asmClient := createASMInterface(jsonSecretValue)
+	secretValueInput := createSecretValueInput(toPtr(valueFrom), nil, nil)
+	nonExistentKey := "nonExistentKey"
+	_, err := GetSecretFromASMWithInput(secretValueInput, asmClient, nonExistentKey)
+	assert.Error(t, err)
+}
+
+func TestGetSecretFromASMWithVersionID(t *testing.T) {
+	asmClient := createASMInterface(secretValue)
+	secretValueInput := createSecretValueInput(toPtr(valueFrom), toPtr(versionID), nil)
+	outSecretValue, err := GetSecretFromASMWithInput(secretValueInput, asmClient, "")
+	if assert.NoError(t, err) {
+		assert.Equal(t, secretValue, outSecretValue)
+	}
+}
+
+func TestGetSecretFromASMWithVersionIDAndStage(t *testing.T) {
+	asmClient := createASMInterface(secretValue)
+	secretValueInput := createSecretValueInput(toPtr(valueFrom), toPtr(versionID), toPtr(versionStage))
+	outSecretValue, err := GetSecretFromASMWithInput(secretValueInput, asmClient, "")
+	if assert.NoError(t, err) {
+		assert.Equal(t, secretValue, outSecretValue)
+	}
+}
+
+func toPtr(input string) *string {
+	if input == "" {
+		return nil
+	}
+	output := input
+	return &output
```

Comment:
You should be able to directly use &input.

Reply:
could not bc they are const values being passed in.

Reply:
Isn't it a variable inside the function?
```
const (
   test = "test"
)

func main() {
	fmt.Println(toPtr(test))
}

func toPtr(input string) *string {
	if input == "" {
		return nil
	}
	return &input
}
```
I tried this in go playground, and it works for me.

Reply:
oops sorry you are right - i forgot that the const were already passed in as a parameter. updated

Reply:
Does the sdk client behave differently between passing the pointer of an empty string vs. passing a nil pointer? If they are different, then your original code is correct